### PR TITLE
Issue-33519. Added logic for grouping alerts by store

### LINKED ISSUE #33519: Template file retrieving should be store scoped
### Preconditions (*)
1. Magento 2.3.6 or lower (maybe greater)

### Steps to reproduce (*)
Setup Multiple Stores/websites in Magento
Create 2 different Stock Alert Email Templates for each store (Marketing - Communications - Email Templates - Add New Template)
Enable Product Alerts and Apply the newly created templates for each Store
Stores - Configuration - Catalog - Catalog - Product Alerts - Stock Alert Email Template - Set the new template created for new Store view scope , "Allow Alert When Product Comes Back in Stock" - Yes
Default Scope: Product Alerts Run Settings - Give the Start time and Save Configuration
Change the Scope of Store View to another Store view and apply other newly created template and save the configuration
Reindex and clear cache
Admin - Add a new Product - Set the Product to Out Of Stock and Save
Front-end - New Store website (for ex: France website) - Sign in as customer - Product Details page - Click on "Notify me when this product is in stock"
Open the same Product in different Store (for ex: Germany website) - Sign in as same customer - Product Details page - Click on "Notify me when this product is in stock"
Admin - Edit the same product and Set the stock status to "In Stock" and Save
Run the Cron > php bin/magento cron:run
Verify emails of both the Stores

### Expected result (*)
1. Each email should have the right template

### Actual result (*)
1. Only the first processed store template is serving

### Workaround
I found out that it's caused by `\Magento\Framework\View\Element\Template\File\Resolver::getTemplateFileName`
In this case, the $key to store the template and avoid multiple times the same process to run is the concatenation of the template identifier with params given by `\Magento\Framework\View\Element\Template`.

In this case i recommend in `\Magento\Framework\View\Element\Template::getTemplateFile` to add in parameters the store id in parameters by this way :

`$params = ['module' => $this->getModuleName()];`
could be
```
$params = [
            'module' => $this->getModuleName(),
            'store_id' => $this->_storeManager->getStore()->getId()
        ];
```
or
```
$store = $this->_storeManager->getStore();
        $params = [
            'module' => $this->getModuleName(),
            'store_id' => $store ? $store->getId() : \Magento\Store\Model\Store::DEFAULT_STORE_ID
        ];

```
---
Please provide [Severity](https://devdocs.magento.com/guides/v2.4/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [x] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by topanok**
### Description (*)
Added logic for grouping alerts by store. The problem was that when sorting collection items, the store ID of the last item of current user is always add to the email object. This fix also fixed the situation with PriceAlerts, because the logic is used by everyone.

### Fixed Issues (if relevant)
1. Fixes magento/magento2#33519

### Manual testing scenarios (*)
1. Setup Multiple Stores/websites in Magento
2. Create 2 different Stock Alert Email Templates for each store (Marketing - Communications - Email Templates - Add New Template)
3. Enable Product Alerts and Apply the newly created templates for each Store
4. Stores - Configuration - Catalog - Catalog - Product Alerts - Stock Alert Email Template - Set the new template created for new Store view scope , "Allow Alert When Product Comes Back in Stock" - Yes
Default Scope: Product Alerts Run Settings - Give the Start time and Save Configuration
Change the Scope of Store View to another Store view and apply other newly created template and save the configuration
5. Reindex and clear cache
6. Admin - Add a new Product - Set the Product to Out Of Stock and Save
7. Front-end - New Store website (for ex: France website) - Sign in as customer - Product Details page - Click on "Notify me when this product is in stock"
8. Open the same Product in different Store (for ex: Germany website) - Sign in as same customer - Product Details page - Click on "Notify me when this product is in stock"
9. Admin - Edit the same product and Set the stock status to "In Stock" and Save
10. Run the Cron > php bin/magento cron:run
11. Verify emails of both the Stores(You should receive two emails from both Stores with different email templates)

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by topanok** [comment]
@magento run all tests

**Comment by topanok** [comment]
@magento run all tests

**Comment by topanok** [comment]
@magento run all tests

**Comment by topanok** [comment]
@magento run all tests

**Comment by topanok** [comment]
@magento run Functional Tests B2B, Functional Tests EE

**Comment by topanok** [comment]
@magento run all tests

**Comment by topanok** [comment]
@magento run Integration Tests, WebAPI Tests

**Comment by topanok** [comment]
@magento run Integration Tests, WebAPI Tests

**Comment by topanok** [comment]
@magento run all tests

**Comment by topanok** [comment]
@magento run Integration Tests, Functional Tests B2B

**Comment by topanok** [comment]
@magento run Functional Tests B2B

**Comment by topanok** [comment]
https://github.com/magento run all tests

**Comment by topanok** [comment]
@magento run Integration Tests, WebAPI Tests, Unit Tests, Functional Tests CE, Functional Tests B2B

**Comment by topanok** [comment]
@magento run all tests

**Comment by topanok** [comment]
@magento run Functional Tests EE, Functional Tests CE, Functional Tests B2B

**Comment by topanok** [comment]
@magento run Functional Tests EE, Functional Tests CE, Functional Tests B2B

**Comment by topanok** [comment]
@magento run all tests

**Comment by topanok** [comment]
@magento run all tests

**Comment by topanok** [comment]
@magento run all tests

**Comment by topanok** [comment]
https://github.com/magento run all tests

**Comment by topanok** [comment]
@magento run all tests

**Comment by topanok** [comment]
@magento run all tests

**Comment by topanok** [comment]
@magento run all tests

**Comment by topanok** [comment]
@magento run all tests

**Comment by topanok** [comment]
@magento run all tests

**Comment by topanok** [comment]
@magento run all tests

**Comment by topanok** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by topanok** [comment]
@magento run Sample Data Tests B2B, Sample Data Tests CE, Sample Data Tests EE

**Review by engcom-Hotel — CHANGES_REQUESTED** [review]
Hello @topanok,

Thanks for the contribution!

Changes looks good to us for this PR, but please cover the changes with some automated tests as part of the [definition of done](https://developer.adobe.com/commerce/contributor/guides/code-contributions/definition-of-done/#definition-of-done-1)

Thanks

**Comment by topanok** [comment]
> engcom-Hotel
I can't run Sample Data Tests B2B
Sample Data Tests CE 
Sample Data Tests EE 
tests.
You can see the message - Failed to run the builds. Please try to re-run them later.

**Comment by engcom-Hotel** [comment]
@topanok you can ignore that for now.

**Comment by engcom-Echo** [comment]
@topanok Thank you for your contribution.
can you please confirm if you are working on above [comment](https://github.com/magento/magento2/pull/35996#pullrequestreview-1430698809).

**Comment by topanok** [comment]
> 

Yes! What do I need to do? You can see that all tests without sample data have been completed.

**Comment by engcom-Echo** [comment]
> > 
> 
> @engcom-Echo Yes! What do I need to do? You can see that all tests without sample data have been completed.

Can you please cover PR changes with appropriate automated test coverage like Unit Test / Integration Test

**Comment by engcom-Echo** [comment]
@topanok can you please confirm if you are working on above [comment](https://github.com/magento/magento2/pull/35996#issuecomment-1561051959)?

**Comment by topanok** [comment]
@engcom-Echo Yes, I'm working on it.

**Comment by engcom-Echo** [comment]
Hi @topanok,

Did you get a chance to work on above [comment](https://github.com/magento/magento2/pull/35996#issuecomment-1567896602)?

**Comment by topanok** [comment]
> Hi @topanok,
> 
> Did you get a chance to work on above [comment](https://github.com/magento/magento2/pull/35996#issuecomment-1567896602)?

Hi @engcom-Echo . A bit busy right now, but I'm working on tests as well

**Comment by engcom-Hotel** [comment]
Hello @topanok,

Thanks for the response!

As you have mentioned that you will be working on the test coverage, we are moving this issue `On Hold`.

Also, I need to mention that, there is an open issue related to triggering emails for stock and price alerts, please refer to the below issue:

https://github.com/magento/magento2/issues/35602

The above issue is already confirmed. The above issue might be connected with this PR.

Thanks

**Comment by topanok** [comment]
Hello @engcom-Echo. I wrote a test, you can see in the last commit

**Comment by topanok** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
moving to `Pending Review` as automated test coverage is added.

**Comment by engcom-Hotel** [comment]
@magento run WebAPI Tests, Functional Tests EE, Functional Tests B2B

**Comment by engcom-Hotel** [comment]
Hello @topanok,

Thanks for adding the tests. But as per my comment here https://github.com/magento/magento2/pull/35996#issuecomment-1611595125, there is an existing issue due to which we are not able to get the Price Alert emails:

https://github.com/magento/magento2/issues/35602

We need to keep this PR `On Hold`. We will pick this up again whenever that issue has been resolved.

Thanks

**Comment by topanok** [comment]
> As per [this](https://github.com/magento/magento2/pull/35996#issuecomment-1621904068), moving to Waiting Related PRs.
> 
> Thank you!

Hello! Any updates?

**Comment by engcom-Hotel** [comment]
Hello @topanok,

As per the JIRA related to the issue https://github.com/magento/magento2/issues/35602, the status is still in implementation.

We will update you accordingly on the status of that issue.

Thanks

**Comment by engcom-Charlie** [comment]
It seems that bot has moved this PR to Changes Requested hence moving it back to Waiting Related PRs.

**Comment by engcom-Charlie** [comment]
It seems that bot has moved this PR to Changes Requested hence moving it back to Waiting Related PRs.

**Comment by andrewbess** [comment]
@magento run all tests

**Comment by topanok** [comment]
@magento run Functional Tests CE, Functional Tests B2B

**Comment by engcom-Hotel** [comment]
As per this issue comment https://github.com/magento/magento2/issues/35602#issuecomment-1928965021, the issue is more reproducible.

Hence moving this PR back to `Pending Review`.

Thanks

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests B2B, Functional Tests EE, Functional Tests CE

File: dev/tests/integration/framework/Magento/TestFramework/Mail/TransportInterfaceMock.php [hunk 1]
```diff
@@ -35,6 +40,7 @@ public function __construct($message = null)
      */
     public function sendMessage()
```
Comment (engcom-Hotel) [inline dev/tests/integration/framework/Magento/TestFramework/Mail/TransportInterfaceMock.php:41]: We can have a return type here as `void` maybe
```suggestion
    public function sendMessage() : void
```

File: dev/tests/integration/framework/Magento/TestFramework/Mail/TransportInterfaceMock.php [hunk 2]
```diff
@@ -48,4 +54,25 @@ public function getMessage()
     {
         return $this->message;
     }
+
+    /**
+     * Returns count of sent messages
+     *
+     * @return int
+     */
+    public function getSentMsgCount()
```
Comment (engcom-Hotel) [inline dev/tests/integration/framework/Magento/TestFramework/Mail/TransportInterfaceMock.php]: Same here

File: dev/tests/integration/framework/Magento/TestFramework/Mail/TransportInterfaceMock.php [hunk 3]
```diff
@@ -48,4 +54,25 @@ public function getMessage()
...
+        return self::$sentMsgCount;
+    }
+
+    /**
+     * Reset amount of sent message
+     *
+     * @param int $count
+     * @return void
+     */
+    public function resetSentMsgCount($count = 0)
```
Comment (engcom-Hotel) [inline dev/tests/integration/framework/Magento/TestFramework/Mail/TransportInterfaceMock.php]: and here

**Review by engcom-Hotel — CHANGES_REQUESTED** [review]
Hello @topanok,

Thanks for the contribution!

Code changes seem good to me, please make the changes below as suggested.

Failed tests seem flaky to me.

Thanks

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests B2B, Functional Tests EE, Integration Tests

**Review by engcom-Hotel — APPROVED** [review]

**Comment by engcom-Delta** [comment]
@topanok, We have tried to reproduce and test this issue but we are not getting any product stock alert email.

Steps to reproduce:

1.Setup Multiple Stores/websites in Magento
2.Create 2 different Stock Alert Email Templates for each store (Marketing - Communications - Email Templates - Add New Template)
3.Enable Product Alerts and Apply the newly created templates for each Store
4.Stores - Configuration - Catalog - Catalog - Product Alerts - Stock Alert Email Template - Set the new template created for new Store view scope , "Allow Alert When Product Comes Back in Stock" - Yes
Default Scope: Product Alerts Run Settings - Give the Start time and Save Configuration
5.Change the Scope of Store View to another Store view and apply other newly created template and save the configuration
6.Reindex and clear cache
7Admin - Add a new Product - Set the Product to Out Of Stock and Save
8.Front-end - New Store website (for ex: France website) - Sign in as customer - Product Details page - Click on "Notify me when this product is in stock"
9.Open the same Product in different Store (for ex: Germany website) - Sign in as same customer - Product Details page - Click on "Notify me when this product is in stock"
10.Admin - Edit the same product and Set the stock status to "In Stock" and Save
11.Run the Cron > php bin/magento cron:run
12.Verify emails of both the Stores

Screenshot 2024-02-28 at 6 49 58 PM
Screenshot 2024-02-28 at 6 50 22 PM
Screenshot 2024-02-28 at 6 51 03 PM
Screenshot 2024-02-28 at 6 51 52 PM
Screenshot 2024-02-28 at 6 52 55 PM
Screenshot 2024-02-28 at 6 53 50 PM
Screenshot 2024-02-28 at 6 54 50 PM
Screenshot 2024-02-28 at 6 55 25 PM
Screenshot 2024-02-28 at 6 55 32 PM
Screenshot 2024-02-28 at 6 55 45 PM
Screenshot 2024-02-28 at 6 55 57 PM
Screenshot 2024-02-28 at 6 56 06 PM
Screenshot 2024-02-28 at 6 57 19 PM
Screenshot 2024-02-28 at 6 56 40 PM
Screenshot 2024-02-28 at 7 08 44 PM


Kindly suggest if something is missing

**Comment by topanok** [comment]
@engcom-Delta Hello! Mail server configured? Can you get any email like password reset?

**Comment by engcom-Delta** [comment]
> @engcom-Delta Hello! Mail server configured? Can you get any email like password reset? And wait until midnight)) Снимок экрана 2024-02-29 в 20 32 32


@topanok, Yes Mail server configured and i am getting password reset email but not the alert once till now

**Comment by topanok** [comment]
@engcom-Delta Perhaps this is related to the time of sending letters, which you have set to midnight, and the one-time start of the cron did not work. Start the cron in constant operation with the `php bin/magento cron:install` command and wait for the letter after midnight or change the sending time to the nearest one. I hope you cleared the config cache after saving?

**Comment by topanok** [comment]
@engcom-Delta I checked the functionality. Everything works if the cron is running. Namely... The time set in the admin is responsible for starting the cron, which adds operations to the queue. After that, the cron works, which performs the scheduled operations. It turns out that the command from the message above (`php bin/magento cron:install`) should fix the situation. But you won't get notifications until midnight. 
If you do not receive anything, then subscribe to the product again, because something may have gone wrong due to the disabled cron. This is partly my fault for writing the wrong command to run kron. Sorry!)

**Comment by engcom-Delta** [comment]
> @engcom-Delta I checked the functionality. Everything works if the cron is running. Namely... The time set in the admin is responsible for starting the cron, which adds operations to the queue. After that, the cron works, which performs the scheduled operations. It turns out that the command from the message above (`php bin/magento cron:install`) should fix the situation. But you won't get notifications until midnight. If you do not receive anything, then subscribe to the product again, because something may have gone wrong due to the disabled cron. This is partly my fault for writing the wrong command to run kron. Sorry!)

@topanok, That's okay and my cron is working fine but we are not getting the alert mails and some more PRs are on hold due to that. May i know in which version of magento it is working for you and more details if any.

Thanks,

**Comment by engcom-Delta** [comment]
@topanok Waiting for your reply on the PR to move ahead.

**Comment by topanok** [comment]
> @topanok Waiting for your reply on the PR to move ahead.

Hello. Tested on the same branch as PR and Magento version dev-2.4-develop. Please show what you have in the `product_alert_stock` table in the marked columns. 
Снимок экрана 2024-03-26 в 20 18 51

**Comment by topanok** [comment]
> @topanok Waiting for your reply on the PR to move ahead.

Hello! I raised the magento instance yesterday to test everything again. Here are the results. As you can see, the letters have arrived and you can see it in the magento database. So I can assume that you have some problems with the environment. Maybe ask someone for help. 

Снимок экрана 2024-03-27 в 10 30 53
Снимок экрана 2024-03-27 в 10 30 30

**Comment by engcom-Delta** [comment]
Hi @topanok , Thanks for your reply. We are also using 2.4 develop. I have checked the table product_alert_stock. Adding the screenshot for the same.

Screenshot 2024-03-27 at 11 54 10 AM

**Comment by topanok** [comment]
> Hi @topanok , Thanks for your reply. We are also using 2.4 develop. I have checked the table product_alert_stock. Adding the screenshot for the same.
> 
> Screenshot 2024-03-27 at 11 54 10 AM

It looks like you don't have cron running or you don't have mail sent configured. As you can see from my previous post, everything was working fine for me. So maybe get someone else to test it. I don't know how to help you.

**Comment by engcom-Delta** [comment]
> > Hi @topanok , Thanks for your reply. We are also using 2.4 develop. I have checked the table product_alert_stock. Adding the screenshot for the same.
> > Screenshot 2024-03-27 at 11 54 10 AM
> 
> It looks like you don't have cron running or you don't have mail sent configured. As you can see from my previous post, everything was working fine for me. So maybe get someone else to test it. I don't know how to help you.


Hi @topanok , As mentioned in the below comment we are not getting product alert email as of now when cron is running and email alert has set , so we have reopened the issue.We will wait till the issue is resolved to test this PR.

https://github.com/magento/magento2/pull/35996#issuecomment-2022289510

**Comment by topanok** [comment]
> > > Привіт@topanok, Дякуємо за Ваш відповідь. Ми також використовуємо 2.4 development. Я перевірив таблицю product_alert_stock. Додавання скріншота для того ж.
> > > Скріншот 27.03.2024 об 11 54 10 ранку
> > 
> > 
> > Схоже, у вас не запущено cron або не налаштовано надсилання пошти. Як ви бачите з мого попереднього допису, у мене все працювало нормально. Тому, можливо, попросіть когось іншого перевірити це. Я не знаю, як вам допомогти.
> 
> Привіт@topanok, Як зазначено в коментарі нижче, наразі ми не отримуємо сповіщення електронною поштою про продукт, коли запущено cron і встановлено сповіщення електронною поштою, тому ми повторно відкрили проблему. Ми зачекаємо, доки проблему не буде вирішено, щоб перевірити цей PR.
> 
> [#35996 (коментар)](https://github.com/magento/magento2/pull/35996#issuecomment-2022289510)

You can wait, but I won't do anything. It's not my problem that you can't test. I'm already fed up with this pull request, which is already 1.5 years old.

**Comment by topanok** [comment]
> > > Hi @topanok , Thanks for your reply. We are also using 2.4 develop. I have checked the table product_alert_stock. Adding the screenshot for the same.
> > > Screenshot 2024-03-27 at 11 54 10 AM
> > 
> > 
> > It looks like you don't have cron running or you don't have mail sent configured. As you can see from my previous post, everything was working fine for me. So maybe get someone else to test it. I don't know how to help you.
> 
> Hi @topanok , As mentioned in the below comment we are not getting product alert email as of now when cron is running and email alert has set , so we have reopened the issue.We will wait till the issue is resolved to test this PR.
> 
> [#35996 (comment)](https://github.com/magento/magento2/pull/35996#issuecomment-2022289510)

You can wait, but I won't do anything. It's not my problem that you can't test. I'm already fed up with this pull request, which is already 1.5 years old.

**Comment by engcom-Hotel** [comment]
Hello @topanok,

The issue that was preventing us from receiving alert emails has been resolved. We can now move forward with this PR. However, we kindly request that you resolve the conflicts.

Thanks

**Comment by engcom-Hotel** [comment]
Hello @topanok,

A gentle reminder to resolve the conflicts on this PR.

Thanks

**Comment by engcom-Hotel** [comment]
Hello @topanok,

We have noticed that this PR has not been updated for a while. Therefore, we are closing it for now.

Please let us know when you are ready to work on it again, and we will be happy to reopen the PR.

Thank you for your valuable contribution!

**Pull request closed without merging**